compiler/src/dotty/tools/dotc/semanticdb/TypeOps.scala

```diff
@@ -0,0 +1,305 @@
+                case _: TermParamRef => s.SingleType(s.Type.Empty, ssym)
+              }
+            case None => // shouldn't happen
+              s.Type.Empty
+          }
+
+        case ThisType(TypeRef(_, desig)) if desig.isInstanceOf[Symbol] =>
+          val ssym = desig.asInstanceOf[Symbol].symbolName
+          s.ThisType(ssym)
+
+        case SuperType(thistpe, supertpe) =>
+          val spre = loop(thistpe.typeSymbol.info)
+          val ssym = supertpe.typeSymbol.symbolName
+          s.SuperType(spre, ssym)
+
+        // val clazzOf = classOf[...]
+        case ConstantType(const) if const.tag == core.Constants.ClazzTag =>
+          loop(const.typeValue)
+
+        case ConstantType(const) =>
+          s.ConstantType(const.toSemanticConst)
+
+        case rt @ RefinedType(parent, name, info) =>
+          // `X { def x: Int; def y: Int }`
+          // RefinedType(
+          //   parent = RefinedType(
+          //     parent = TypeRef(..., X)
+          //     ...
+          //   )
+          //   refinedName = x
+          //   refinedInfo = TypeRef(..., Int)
+          // )
+          type RefinedInfo = (core.Names.Name, Type)
+          def flatten(tpe: Type, acc: List[RefinedInfo]): (Type, List[RefinedInfo]) = tpe match {
+            case RefinedType(parent, name, info) =>
+              flatten(parent, acc :+ (name, info))
+            case _ =>
+              (tpe, acc)
+          }
+
+          // flatten parent types to list
+          // e.g. `X with Y with Z { refined }`
+          // RefinedType(parent = AndType(X, AndType(Y, Z)), ...)
+          // => List(X, Y, Z)
+          def flattenParent(parent: Type): List[s.Type] = parent match {
+            case AndType(tp1, tp2) =>
+              flattenParent(tp1) ++ flattenParent(tp2)
+            case _ => List(loop(parent))
+          }
+
+          val (parent, refinedInfos) = flatten(rt, List.empty)
+          val stpe = s.WithType(flattenParent(parent))
+
+          val decls = refinedInfos.flatMap { (name, _) =>
+            refinementSymtab.get((rt, name))
+          }
+          val sdecls = decls.sscope(using LinkMode.HardlinkChildren)
+          s.StructuralType(stpe, Some(sdecls))
+
+        case rec: RecType =>
+          loop(rec.parent) // should be handled as RefinedType
+
+        // repeated params: e.g. `Int*`, which is the syntax sugar of
+        // `Seq[Int] @Repeated` (or `Array[Int] @Repeated`)
+        // See: Desugar.scala and TypeApplications.scala
+        case AnnotatedType(AppliedType(_, targs), annot)
+          if (annot matches defn.RepeatedAnnot) && (targs.length == 1) =>
+          val stpe = loop(targs(0))
+          s.RepeatedType(stpe)
+
+        case ann: AnnotatedType if ann.annot.symbol.info.isInstanceOf[ClassInfo] =>
+          def flatten(tpe: Type, annots: List[Annotation]): (Type, List[Annotation]) = tpe match
+            case AnnotatedType(parent, annot) if annot.symbol.info.isInstanceOf[ClassInfo] =>
```

Comment:
might want to be careful of type aliases used as annotations here, which won't have `ClassInfo` as their info, maybe `annot.symbol` takes care of this

Reply:
Judging from the `Annotations.symbol` https://github.com/lampepfl/dotty/blob/601184787508d8fd655f1dc4e48ed19c418fa710/compiler/src/dotty/tools/dotc/core/Annotations.scala#L13-L15 I guess we can expect there should only be `ClassInfo`.
Since I tried but couldn't find a counter-example for this, maybe we can go with it, and fix it in the future when we find a example here.

Reply:
I confirmed with testing it does not matter if the type is aliased, however it does not work for nested annotated types, e.g.
```scala
type paramAlias = param
type paramRec = param @param
type AnnotatedType = Int @param
type AnnotatedType2 = Int @paramAlias
type AnnotatedType3 = Int @(param @param)
type AnnotatedType4 = Int @paramRec
```
In all these cases `AnnotatedType<N>` info is `Int @param` but scala 2 would have `Int @param @param`

Reply:
I believe this still needs resolving